onnxruntime/core/providers/cpu/signal/dft.cc

```diff
@@ -181,50 +181,130 @@ static Status fft_radix2(OpKernelContext* /*ctx*/, const Tensor* X, Tensor* Y, s
+  }
 
   // Get data
   auto* X_data = const_cast<U*>(reinterpret_cast<const U*>(X->DataRaw())) + X_offset;
   auto* Y_data = reinterpret_cast<std::complex<T>*>(Y->MutableDataRaw()) + Y_offset;
-
   U* window_data = nullptr;
   if (window) {
     window_data = const_cast<U*>(reinterpret_cast<const U*>(window->DataRaw()));
   }
 
-  auto angular_velocity = compute_angular_velocity<T>(dft_length, inverse);
+  auto a = onnxruntime::Tensor(X->DataType(), dft_input_shape, alloc);
+  auto a_fft = onnxruntime::Tensor(Y->DataType(), dft_input_shape, alloc);
+  std::complex<T>* a_data = reinterpret_cast<std::complex<T>*>(a.MutableDataRaw());
+  std::complex<T>* a_fft_data = reinterpret_cast<std::complex<T>*>(a_fft.MutableDataRaw());
+  std::complex<T>* b_fft_data = reinterpret_cast<std::complex<T>*>(b_fft.MutableDataRaw());
+  std::complex<T>* chirp_data = reinterpret_cast<std::complex<T>*>(chirp.MutableDataRaw());
+  memset(reinterpret_cast<void*>(a_data), 0, a.SizeInBytes());
+
+  const auto& X_shape = X->Shape();
+  size_t number_of_samples = static_cast<size_t>(X_shape[onnxruntime::narrow<size_t>(axis)]);
+
+  // Prepare "a" signal
+  for (size_t n = 0; n < number_of_samples; n++) {
+    std::complex<T>& a_n = *(a_data + n);
+    std::complex<T>& chirp_n = *(chirp_data + n);
+    auto window_n = window_data ? *(window_data + n) : 1;
+    a_n = *(X_data + n * X_stride);  // input
+    a_n *= window_n;
+    a_n *= chirp_n;
+  }
+
+  // Forward FFT radix2 for the "a" signal
+  ORT_RETURN_IF_ERROR((fft_radix2<T, std::complex<T>>(ctx, &a, &a_fft, 0, 1, 0, 1, 1, M, nullptr,
+                                        false, inverse, V, temp_output)));
+
+  for (size_t i = 0; i < M; i++) {
+    std::complex<T>& a_i = *(a_fft_data + i);
+    std::complex<T>& b_i = *(b_fft_data + i);
+    a_i *= b_i;
+  }
+
+  // Inverse FFT radix2 for the "a" signal
+  ORT_RETURN_IF_ERROR((fft_radix2<T, std::complex<T>>(ctx, &a_fft, &a, 0, 1, 0, 1, 1, M, nullptr,
+                                        false, !inverse, V, temp_output)));
+  const auto& Y_shape = Y->Shape();
+  size_t dft_output_size = static_cast<size_t>(Y_shape[onnxruntime::narrow<size_t>(axis)]);
 
   for (size_t i = 0; i < dft_output_size; i++) {
+    std::complex<T>& chirp_i = *(chirp_data + i);
     std::complex<T>& out = *(Y_data + i * Y_stride);
-    out.real(0);
-    out.imag(0);
-
-    for (size_t j = 0; j < dft_length; j++) {  // vectorize over this loop
-      auto exponential = compute_exponential(i * j, angular_velocity);
-      auto window_element = window_data ? *(window_data + j) : 1;
-      auto x = (j < number_of_samples) ? *(X_data + j * X_stride) : 0;
-      auto element = x * window_element;
-      out += exponential * element;
-    }
-
-    if (inverse) {
-      out /= static_cast<T>(dft_length);
+    std::complex<T>& c_i = *(a_data + i);
+    if (i > 0) {
+      c_i = *(a_data + M - i); // Why reverse the output?
```

Comment:
Did you mean to leave that comment in there?

Reply:
Updated comment.